src/avro_json_decoder.erl

```diff
@@ -266,68 +275,70 @@ type_from_name(_)                 -> undefined.
 %%% Values parsing
 %%%===================================================================
 
-parse_value(null, Type, _ExtractFun) when ?AVRO_IS_NULL_TYPE(Type) ->
-  avro_primitive:null();
+parse_value(Value, Type, ExtractFun) ->
+  parse(Value, Type, ExtractFun, _IsWrapped = true).
 
-parse_value(V, Type, _ExtractFun) when ?AVRO_IS_BOOLEAN_TYPE(Type) andalso
-                                       is_boolean(V) ->
+parse(Value, TypeName, ExtractFun, IsWrapped) when is_list(TypeName) ->
+  %% Type is defined by its name
+  Type = ExtractFun(TypeName),
+  parse(Value, Type, ExtractFun, IsWrapped);
+parse(Value, Type, _ExtractFun, IsWrapped) when ?AVRO_IS_PRIMITIVE_TYPE(Type) ->
+  WrappedValue = parse_prim(Value, Type),
```

Comment:
So here we wrap first and then unwrap when IsWrapped =:= false.

What if parse_prim would return an unwrapped value + Fun to wrap it? Then call Fun when IsWrapped =:= true.